javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -1014,6 +1043,201 @@ object Formatter {
+        val unscaledValue = s.substring(0, significandEnd)
+        val scale = -e
+        new Decimal(negative, unscaledValue, scale)
+      } else {
+        // There is a '.'; there can be leading 0's, which we must remove
+        val digits = s.substring(0, dotPos) + s.substring(dotPos + 1, significandEnd)
+        val digitsLen = digits.length()
+        var i = 0
+        while (i < digitsLen && digits.charAt(i) == '0')
+          i += 1
+        val unscaledValue = digits.substring(i)
+        val scale = -e + (significandEnd - (dotPos + 1))
+        new Decimal(negative, unscaledValue, scale)
+      }
+    }
+  }
+
+  /** A decimal representation of a number.
+   *
+   *  An instance of this class represents the number whose absolute value is
+   *  `unscaledValue × 10^(-scale)`, and that is negative iff `negative` is
+   *  true.
+   *
+   *  The `unscaledValue` is stored as a String of decimal digits, i.e.,
+   *  characters in the range ['0', '9'], expressed in base 10. Leading 0's are
+   *  *not* valid.
+   *
+   *  As an exception, a zero value is represented by an `unscaledValue` of
+   *  `"0"`. The scale of zero value is always 0.
+   *
+   *  `Decimal` is similar to `BigDecimal`, with some differences:
+   *
+   *  - `Decimal` distinguishes +0 from -0.
+   *  - The unscaled value of `Decimal` is stored in base 10.
+   *
+   *  The methods it exposes have the same meaning as for BigDecimal, with the
+   *  only rounding mode being HALF_UP.
+   */
+  private final class Decimal(val negative: Boolean, val unscaledValue: String,
+      val scale: Int) {
+
+    def isZero: Boolean = unscaledValue == "0"
+
+    /** The number of digits in the unscaled value.
+     *
+     *  The precision of a zero value is 1.
+     */
+    def precision: Int = unscaledValue.length()
+
+    /** Rounds the number so that it has at most the given precision, i.e., at
+     *  most the given number of digits in its `unscaledValue`.
+     *
+     *  The given `precision` must be greater than 0.
+     */
+    def round(precision: Int): Decimal = {
+      if (precision <= 0)
+        throw new AssertionError("Decimal.round() called with non-positive precision")
+
+      if (precision >= unscaledValue.length()) {
+        // no rounding needed (this branch is always taken for zero values)
+        this
+      } else {
+        val diff = unscaledValue.length() - precision
+        val rounded = roundDigitsAtPos(unscaledValue, precision)
+        val roundedLen = rounded.length()
+
+        if (roundedLen > precision) {
+          // the carry expanded all the way to a new digit
+          if (rounded.charAt(roundedLen - 1) != '0')
+            throw new AssertionError(s"$this.round($precision) caused an invalid rounding")
+          val amendedDigits = rounded.substring(0, roundedLen - 1)
+          val newScale = scale - diff - 1
+          new Decimal(negative, amendedDigits, newScale)
```

Comment:
Why do we do the whole string un-mangling here? IIUC `round` doesn't give any guarantees about the resulting scale (and doesn't need to).

Reply:
It does give a guarantee about the `precision`, though. If we don't do this, it can return a `Decimal` whose `precision` (`unscaledValue.length()`) is > the `precision` parameter, breaking the spec of this method.

That spec is useful in `scientificNotation` because it means we only ever need to *pad* with missing zeros at the end, and not to truncate the `unscaledValue`.